File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn is-shout
  "Is it all caps?"
  [s]
  (and
    (= (str/upper-case s) s) 
    (< 0 (count (re-seq #"[A-Z]" s)))))

(defn response-for
  "Talk to Bob"
  [s]
  (cond
    ;; Handle silence
    (str/blank? s) "Fine. Be that way!"
    ;; Handle yelling
    (is-shout s) "Whoa, chill out!"
    ;; Handle questions 
    (str/ends-with? s "?") "Sure."
    ;; Otherwise
    :else "Whatever.")) 

File: haskell/bob/Bob.hs
module Bob where

import qualified Data.Text as T
import qualified Data.Char as C

isShout :: T.Text -> Bool
isShout s = (T.toUpper s == s) &&
            (T.length $ T.filter C.isAlpha s) > 0

isQuestion :: T.Text -> Bool
isQuestion s = T.last s == '?'

isSilence :: T.Text -> Bool
isSilence s = T.null $ T.strip s

responseFor :: String -> String
responseFor q
    | isSilence s  = "Fine. Be that way!"
    | isShout s    = "Whoa, chill out!"
    | isQuestion s = "Sure."
    | otherwise    = "Whatever."
    where
        s = T.pack q